state/pool.go

```diff
@@ -21,47 +21,69 @@ import (
 
 var errPoolClosed = errors.New("pool closed")
 
-// NewStatePool returns a new StatePool instance. It takes a State
-// connected to the system (controller model).
-func NewStatePool(systemState *State) *StatePool {
-	pool := &StatePool{
-		systemState: systemState,
-		pool:        make(map[string]*PoolItem),
-		hub:         pubsub.NewSimpleHub(nil),
+// PoolHelper describes methods for working with a pool-supplied state.
+type PoolHelper interface {
+	Release() bool
+	Annotate(string)
+}
+
+// PooledState is a wrapper for a State reference, indicating that it is
+// managed by a pool.
+type PooledState struct {
+	*State
+
+	pool          *StatePool
+	modelUUID     string
+	isSystemState bool
+	released      bool
+	itemKey       uint64
+}
+
```

Comment:
Do you know the nil pointer check for checking types?

var _ PoolHelper = (*PooledState)(nil)
Has the property that it gives a direct error about PooledState not implementing PoolHelper if there are any problems, which tends to help localize the issue, rather than getting it as part of a function that needs interface A which isn't implemented.
They essentially do the same thing, but this is much more concrete for developers. (I definitely intend to implement this interface).

Reply:
Nice; I have seen instances of this in the code. Will add.